stl/inc/bit

```diff
@@ -94,20 +95,128 @@ _NODISCARD constexpr _Ty rotr(const _Ty _Val, const int _Rotation) noexcept {
+    constexpr int _Digits   = numeric_limits<_Ty>::digits;
+    // lzcnt (when it doesn't fall back to bsr) is defined correctly for zero
+    if (!_Have_lzcnt && _Val == 0) {
         return _Digits;
     }
+    // we use lzcnt (actually bsr if lzcnt is not supported) for all cases now we know
+    // we're not zero. We can do this because lzcnt and bsr share the same instruction
+    // encoding.
+    if constexpr (_Digits <= 16) {
+        return static_cast<int>(__lzcnt16(_Val) - (16 - _Digits));
+    } else if constexpr (_Digits <= 32) {
+        return static_cast<int>(__lzcnt(_Val) - (32 - _Digits));
+    } else {
+#ifdef _M_IX86
+        if (_Have_lzcnt) {
+            return (__lzcnt(_Val >> 32) + __lzcnt(static_cast<unsigned int>(_Val)) - (64 - _Digits);
+        } else {
+            unsigned int _High = _Val >> 32;
+            unsigned int _Low  = static_cast<unsigned int>(_Val);
+            int _Result        = 0;
+            _Result += _High ? __lzcnt(_High) : 0;
+            _Result += _Low ? __lzcnt(_Low) : 0;
+            return result;
+        }
+#else
+        return static_cast<int>(__lzcnt64(_Val) - (64 - _Digits));
+#endif
+    }
+    // note we don't need to call a fallback here because
+    // all supported intel processors at least have bsr/bsf
+}
 
-    if constexpr (sizeof(_Ty) <= sizeof(unsigned int)) {
-        return __builtin_clz(_Val) - (numeric_limits<unsigned int>::digits - _Digits);
+template <typename _Ty>
+inline int _Checked_x86_x86_64_popcount(_Ty _Val) {
+    static bool _Have_popcnt = __Is_popcnt_supported();
+    constexpr int _Digits    = numeric_limits<_Ty>::digits;
+    if (_Have_popcnt) {
+        if constexpr (_Digits <= 16) {
+            return static_cast<int>(__popcnt16(_Val));
+        } else if constexpr (_Digits <= 32) {
+            return static_cast<int>(__popcnt(_Val));
+        } else {
+#ifdef _M_IX86
+            return static_cast<int>(__popcnt(_Val >> 32) + __popcnt(static_cast<unsigned int>(_Val)));
+#else
+            return static_cast<int>(__popcnt64(_Val));
+#endif
+        }
     } else {
-        return __builtin_clzll(_Val) - (numeric_limits<unsigned long long>::digits - _Digits);
+        return _Popcount_helper(_Val);
     }
 }
+#endif // defined(_M_IX86) || defined(_M_X64)
 
+
+#if defined(_M_ARM) || defined(_M_ARM64)
+template <typename _Ty>
+inline int _Checked_arm_countl_zero(_Ty _Val) {
+    constexpr int _Digits = numeric_limits<_Ty>::digits;
+    if (_Val == 0) {
+        return _Digits;
+    }
+    if constexpr (_Digits <= 32) {
+        return _CountLeadingZeros(_Val);
+    } else {
+        return _CountLeadingZeros64(_Val);
+    }
+}
+#endif // defined(_M_ARM) || defined(_M_ARM64)
+
+template <class _Ty, enable_if_t<_Is_standard_unsigned_integer<_Ty>, int> _Enabled>
+_NODISCARD constexpr int countl_zero(const _Ty _Val) noexcept {
+    if (is_constant_evaluated()) {
+        return _Countl_zero_helper(_Val);

```

Comment:
This is defined on line 1030; I suspect that this doesn't obey the two-phase lookup rules (since it hasn't been declared at this point, and isn't being found through ADL).